---
title: "Demulticoder 16S Mothur SOP Validation"
output: html_document
date: "2024-12-24"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(phyloseq); packageVersion("phyloseq")
library(Biostrings); packageVersion("Biostrings")
library(ggplot2); packageVersion("ggplot2")
library(dada2); packageVersion("dada2")
devtools::install_github("grunwaldlab/demulticoder", force=TRUE)
library("demulticoder"); packageVersion("demulticoder")

seed <- 1
set.seed(seed)
```

## Demonstration of how to use demulticoder on a dataset that is actually three separate datasets (RPS10, ITS, and 16S) at once

Input metadata and primerinfo_params files are in data folder  

The only required columns are the first with sample names, and the second with the primer name/barcode used. The subsequent columns are user-specific columns for downstream steps
*metadata.csv file*
```{r, echo=FALSE}
# First, create your data frame
data <- data.frame(
  sample_name = c("F3D0", "F3D1", "F3D141", "F3D142", "F3D143", "F3D144", "F3D145", 
                  "F3D146", "F3D147", "F3D148", "F3D149", "F3D150", "F3D2", "F3D3", 
                  "F3D5", "F3D6", "F3D7", "F3D8", "F3D9", "Mock"),
  primer_name = rep("r16S", 20),
  Day = c(0, 1, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 2, 3, 5, 6, 7, 8, 9, NA),
  When = c(rep("Early", 2), rep("Late", 10), rep("Early", 7), NA)
)

# Now use knitr::kable() to create the table
knitr::kable(data, row.names = FALSE)
```

I then included the necessary second file with the name of the barcode selected, primer sequences, and the optional DADA2 parameter options. I referenced the DADA2 [tutorial](https://benjjneb.github.io/dada2/tutorial.html) to select the proper parameter options.  

Note, primers were already trimmed from reads, but just to be certain, I included the Earth Microbiome primers described [here](https://earthmicrobiome.org/protocols-and-standards/16s/), and a few primer sequences were still found within a small number of reads. 
*primerinfo_params.csv*
```{r, echo=FALSE}
# Create the data frame
primer_info <- data.frame(
  primer_name = "r16S",
  forward = "GTGYCAGCMGCCGCGGTAA",
  reverse = "GGACTACNVGGGTWTCTAAT",
  already_trimmed = TRUE,
  minCutadaptlength = 50,
  multithread = TRUE,
  verbose = TRUE,
  maxN = 0,
  maxEE_forward = 2,
  maxEE_reverse = 2,
  truncLen_forward = 240,
  truncLen_reverse = 160,
  truncQ = 2,
  minLen = 20,
  maxLen = Inf,
  minQ = 0,
  trimLeft = 0,
  trimRight = 0,
  rm.lowcomplex = 0,
  minOverlap = 12,
  maxMismatch = 0,
  min_asv_length = 50
)

# Use knitr::kable() to create the table
knitr::kable(primer_info, row.names = FALSE)
```

### Step 1-Remove N's and create directory structure for downstream steps
```{r prepare reads, fig.height=6, fig.width=6}
outputs<-prepare_reads(
  data_directory = "~/benchmark_demulticoder/mothur_16S_sop/data", 
  output_directory = "~/benchmark_demulticoder/mothur_16S_sop/outputs_demulticoder",
  tempdir_path = "~/benchmark_demulticoder/mothur_16S_sop/temp",
  tempdir_id = "SOP_run",
  overwrite_existing = FALSE)
```

### Step 2-Run Cutadapt to remove primers and then trim reads with DADA2 filterAndTrim function 
```{r Remove primers and trim reads, fig.height=6, fig.width=6}
cut_trim(
  outputs,
  cutadapt_path="/opt/homebrew/bin/cutadapt",
  overwrite_existing = FALSE)
```

### Step 3-Core ASV inference step
```{r ASV inference, fig.height=6, fig.width=6}
make_asv_abund_matrix(
  outputs,
  overwrite_existing = FALSE)
```

### Step 4-Assign taxonomy step
```{r assign taxonomy step, fig.height=6, fig.width=6}
assign_tax(
  outputs,
  asv_abund_matrix,
  db_16S="silva_nr99_v138.2_toSpecies_trainset.fa.gz",
  retrieve_files=TRUE,
  overwrite_existing=FALSE)
```

### Step 5-convert asv matrix to taxmap and phyloseq objects with one function
```{r convert matrix to other formats, fig.height=6, fig.width=6}
objs<-convert_asv_matrix_to_objs(outputs, save_outputs=TRUE, overwrite_existing = TRUE)
```

### Step 6-evaluate accuracy using mock community, as shown in dada2 tutorial
```{r examine accuracy, fig.height=6, fig.width=6}
tax_matrix<-read.csv("~/benchmark_demulticoder/mothur_16S_sop/outputs_demulticoder/final_asv_abundance_matrix_r16S.csv")

unqs.mock <- tax_matrix[, c(2, which(colnames(tax_matrix) == "Mock_r16S"))]

unqs.mock <- unqs.mock[unqs.mock$Mock_r16S != 0,]

cat("DADA2 inferred", nrow(unqs.mock), "sample sequences present in the Mock community.\n")

mock.ref <- dada2::getSequences(file.path("mothur_16S_sop/data", "HMP_MOCK.v35.fasta"))
match.ref <- sum(sapply(unqs.mock$sequence, function(x) any(grepl(x, mock.ref))))
cat("Of those,", sum(match.ref), "were exact matches to the expected reference sequences.\n")
```

When looking at mock community sample, we were able to extract 20 bacterial sequences with 0% mismatch, and this matched what was described previously. 

### Step 7-Follow-up work using phyloseq to do side-by-side comparison with dada2 example and to examine alpha diversity results
```{r phyloseq alpha divresity analysis, fig.height=6, fig.width=6}
#Let's prune mock community sample
objs$phyloseq_r16S <- phyloseq::prune_samples(phyloseq::sample_names(objs$phyloseq_r16S) != "Mock_r16S", objs$phyloseq_r16S) # Remove mock sample
alpha_div<-phyloseq::plot_richness(objs$phyloseq_r16S, x="Day", measures=c("Shannon", "Simpson"), color="When")

#Let's save a figure of the alpha diversity results
ggsave("~/benchmark_demulticoder/mothur_16S_sop/figures/alpha_diversity.pdf", alpha_div, width=4, height=3)
```

### Step 8-Examine ordination plots as additional point of comparison with DADA2 tutorial
```{r phyloseq beta diversity analysis, fig.height=6, fig.width=6}
# Transform data to proportions as appropriate for Bray-Curtis distances
ps.prop <- phyloseq::transform_sample_counts(objs$phyloseq_r16S, function(otu) otu/sum(otu))
ord.nmds.bray <- phyloseq::ordinate(ps.prop, method="NMDS", distance="bray")
beta_div <- phyloseq::plot_ordination(ps.prop, ord.nmds.bray, color="When")

ggsave("~/benchmark_demulticoder/mothur_16S_sop/figures/beta_diversity.pdf", beta_div, width=4.5, height=3)
```

### Step 9-Let's look at what the top 20 taxa are in the early vs. late samples  time points, as shown in the dada2 tutorial
```{r phyloseq community analysis, fig.height=5, fig.width=8}
top20 <- names(sort(phyloseq::taxa_sums(objs$phyloseq_r16S), decreasing=TRUE))[1:20]
ps.top20 <- phyloseq::transform_sample_counts(objs$phyloseq_r16S, function(OTU) OTU/sum(OTU))
ps.top20 <- phyloseq::prune_taxa(top20, ps.top20)
top20_phylo<-phyloseq::plot_bar(ps.top20, x="Day", fill="Family") + ggplot2::facet_wrap(~When, scales="free_x")
top20<-ggsave("~/benchmark_demulticoder/mothur_16S_sop/figures/top20_taxa.pdf", top20_phylo, width=4.5, height=3)
```

### Let's facet the pdf files 
```{r facet pdf files, fig.height=6, fig.width=6}
# Create text grobs for the labels
label_a <- textGrob("A. Alpha diversity", x = unit(0, "npc"), just = "left", gp = gpar(fontsize = 18, fontface = "bold"))
label_b <- textGrob("B. Beta diversity", x = unit(0, "npc"), just = "left", gp = gpar(fontsize = 18, fontface = "bold"))
label_c <- textGrob("C. Top taxa", x = unit(0, "npc"), just = "left", gp = gpar(fontsize =18, fontface = "bold"))

# Combine the plots with labels
combined_plot <- gridExtra::grid.arrange(
  label_a, alpha_div,
  label_b, beta_div,
  label_c, top20_phylo,
  ncol = 1,
  nrow = 6,
  heights = c(0.1, 1, 0.1, 1, 0.1, 1)
)

# Save the combined plot as a PDF
ggsave("~/benchmark_demulticoder/mothur_16S_sop/figures/faceted_plots.pdf", combined_plot, width = 6, height = 10)
ggsave("~/benchmark_demulticoder/mothur_16S_sop/figures/faceted_plots.svg", combined_plot, width = 6, height = 10)
```

<<<<<<< HEAD
### Let's collect info on R configuration and associated package versions that are downloaded
=======


>>>>>>> parent of 49520ef (update files)
```{r}
sessioninfo::session_info()
```