---
title: "Cophylo analysis-IQ-TREE2"
output: html_document
date: "2024-12-10"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ape)
library(phytools)
library(phangorn)
library(ggtree)
library(ggplot2)
library(grid)
library(markdown)
library(RColorBrewer)
library(fuzzyjoin)
library(stringr)

knitr::opts_knit$set(root.dir = "~/benchmark_demulticoder")
seed <- 1
set.seed(seed)
```

### Read in the multi-fasta alignments to make neighbor-joining trees
```{r, read in trees}
cox1_aln <- ape::read.FASTA("rps10_cox1_phylogeny/data/cox1_mafft2_aligned_auto_ugenetrim.fasta")
rps10_aln <- ape::read.FASTA("rps10_cox1_phylogeny/data/rps10_mafft2_aligned_auto_ugenetrim.fasta")

datasettable <- read.table("rps10_cox1_phylogeny/data/phyloinfo_cox1rps10.txt", sep = "\t", header=T, stringsAsFactors = F, quote = "")

assoctable <- read.table("rps10_cox1_phylogeny/data/assoc_table.txt", sep = "\t", header=T, stringsAsFactors = F, quote = "")
```


### Make neighbor-joining trees
```{r, make trees}
cox1 <- ape::nj(dist.dna(cox1_aln))
cox1 <- midpoint(cox1)

rps10 <- ape::nj(dist.dna(rps10_aln))
rps10 <- midpoint(rps10)
```

###Tip names should already match, but let's check
```{r, check tip names}
all(cox1$tip.label %in% rps10$tip.label)

cox1_tips <- cox1$tip.label
rps10_tips <- rps10$tip.label

cox1_not_in_rps10 <- setdiff(cox1, rps10_tips)
rps10_not_in_cox1 <- setdiff(rps10_tips, cox1_tips)

cat("Tip labels in cox1 but not in rps10:\n")
print(cox1_not_in_rps10)
cat("\nTip labels in rps10 but not in cox1:\n")
print(rps10_not_in_cox1)


#Check for other set of trees
all(cox1$tip.label %in% rps10$tip.label)

cox1_tips <- cox1$tip.label
rps10_tips <- rps10$tip.label

cox1_not_in_rps10 <- setdiff(cox1, rps10_tips)
rps10_not_in_cox1 <- setdiff(rps10_tips, cox1_tips)

cat("Tip labels in cox1 but not in rps10:\n")
print(cox1_not_in_rps10)
cat("\nTip labels in rps10 but not in cox1:\n")
print(rps10_not_in_cox1)


all.equal(sort(cox1$tip.label), sort(rps10$tip.label))
```

### Check that tip labels match the metadata sheet
```{r tip label checks}
cox1_tips <- cox1$tip.label
rps10_tips <- rps10$tip.label

cox1_not_in_dataset <- setdiff(cox1_tips, datasettable$strain)
rps10_not_in_dataset <- setdiff(rps10_tips, datasettable$strain)
```


### Make cophylogeny
### TODO make sure colors match
```{r, make cophylogeny}
# Define genus colors
#genus_colors <- setNames(c("#56B4E9", "#E69F00","#CC79A7", "#F0E442"), #unique(datasettable$genus))

genus_colors <- setNames(c("black", "black","black", "black"), unique(datasettable$genus))

# Function to color the tips based on the dataset
color_tips <- function(tree, dataset, genus_colors) {
  tip_colors <- genus_colors[dataset$genus[match(tree$tip.label, dataset$strain)]]
  return(tip_colors)
}

# Make association table
cox1_tip_colors <- color_tips(cox1, datasettable, genus_colors)
rps10_tip_colors <- color_tips(rps10, datasettable, genus_colors)

# Debugging: Print out the mappings to check for correctness
print("cox1 tip labels and colors:")
print(data.frame(tip.label = cox1$tip.label, color = cox1_tip_colors))

print("rps10 tip labels and colors:")
print(data.frame(tip.label = rps10$tip.label, color = rps10_tip_colors))

# Check if any NA values are present in the tip colors
if (any(is.na(cox1_tip_colors))) {
  warning("NA values found in cox1_tip_colors")
}
if (any(is.na(rps10_tip_colors))) {
  warning("NA values found in rps10_tip_colors")
}

# Create cophylo object
obj <- phytools::cophylo(multi2di(cox1), multi2di(rps10), assoc=assoctable, rotate.multi = TRUE)

# Plot the cophylogeny
pdf("rps10_cox1_phylogeny/figures/figure_cox_vs_rps10_unique_auto_nj_black.pdf", width = 8.5, height = 11, useDingbats = FALSE)
plot(obj, tip.len = 0.01, fsize = 0.20, scale.bar = c(0.2, 0.2), pts = FALSE, link.lwd=1, link.lty="solid", link.col=make.transparent("black", 0.4))
#title(main = "Cophylogeny", line = -1)

# Add tip labels with colors
tiplabels.cophylo(pch = 21, bg = cox1_tip_colors, cex = 0.4, which = "left")
tiplabels.cophylo(pch = 21, bg = rps10_tip_colors, cex = 0.4, which = "right")

dev.off()
```

```{r session info}
sessioninfo::session_info()
```